gem/optimise.py

```diff
@@ -413,3 +474,403 @@ def aggressive_unroll(expression):
     expression, = unroll_indexsum((expression,), max_extent=numpy.inf)
     expression, = remove_componenttensors((expression,))
     return expression
+
+
+@singledispatch
+def count_flop_node(node):
+    """Count number of flops at a particular gem node, without recursing
+    into childrens"""
+    raise AssertionError("cannot handle type %s" % type(node))
+
+
+@count_flop_node.register(Constant)
+@count_flop_node.register(Terminal)
+@count_flop_node.register(Indexed)
+@count_flop_node.register(Variable)
+@count_flop_node.register(ListTensor)
+@count_flop_node.register(FlexiblyIndexed)
+@count_flop_node.register(IndexSum)
+@count_flop_node.register(LogicalNot)
+@count_flop_node.register(LogicalAnd)
+@count_flop_node.register(LogicalOr)
+@count_flop_node.register(Conditional)
```

Comment:
Yeah... kind of want it to fail loudly for now in case I missed some types, but perhaps should change later.